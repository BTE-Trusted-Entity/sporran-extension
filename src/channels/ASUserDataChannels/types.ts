import { ICredential, KiltAddress } from '@kiltprotocol/types';
import { HexString } from '@polkadot/util/types';

import { DAppName } from '../AccessChannels/DAppName';
import { Origin } from '../AccessChannels/Origin';

export type ASUserDataInput = DAppName & {
  submitter: KiltAddress;
};

export type ASUserDataOriginInput = ASUserDataInput & Origin;

export interface ASUserDataOutput {
  createDidExtrinsic: HexString;
  credential: ICredential;
}
