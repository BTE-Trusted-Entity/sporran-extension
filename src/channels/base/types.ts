export interface ErrorFirstCallback<Output> {
  (error: null, output: Output): void;

  (error: Error): void;
}
