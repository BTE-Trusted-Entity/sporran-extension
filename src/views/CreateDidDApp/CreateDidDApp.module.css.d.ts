export const __esModule: true;
export const container: string;
export const heading: string;
export const buttonsLine: string;
export const details: string;
export const label: string;
export const origin: string;
export const reject: string;
export const submit: string;
export const errorTooltip: string;
