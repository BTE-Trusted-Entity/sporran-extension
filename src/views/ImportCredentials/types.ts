import { KiltAddress } from '@kiltprotocol/sdk-js';

export interface Import {
  fileName: string;
}

export interface SuccessfulImport extends Import {
  identityAddress: KiltAddress;
}

export interface FailedImport extends Import {
  error: 'invalid' | 'orphaned' | string;
}
