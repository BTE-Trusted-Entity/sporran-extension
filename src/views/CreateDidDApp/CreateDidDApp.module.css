.container {
  composes: container from '../../components/View/View.module.css';
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
  margin-bottom: calc(31rem / 16);
}

.buttonsLine {
  composes: buttonsLine from '../../components/Button/Button.module.css';
}

.details {
  margin-top: 1.5rem;
  width: 100%;
}

.label {
  composes: smallText from '../../components/Typography/Typography.module.css';
  width: 100%;
  margin: 0;
  text-align: left;
}

.origin {
  composes: inputText from '../../components/Typography/Typography.module.css';
  width: 100%;
  margin: 0.5rem 0;
  text-align: left;
}

.reject {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
}

.submit {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}

.errorTooltip {
  composes: tooltip from '../../components/ErrorTooltip/ErrorTooltip.module.css';
  top: 2.5rem;
  left: 75%;
}
