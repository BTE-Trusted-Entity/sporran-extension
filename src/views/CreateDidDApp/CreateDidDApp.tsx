import { FormEvent, useCallback } from 'react';
import { browser } from 'webextension-polyfill-ts';
import { Did } from '@kiltprotocol/sdk-js';

import * as styles from './CreateDidDApp.module.css';

import { CreateDidOriginInput } from '../../channels/CreateDidChannels/types';
import {
  PasswordField,
  usePasswordField,
} from '../../components/PasswordField/PasswordField';
import { isFullDid, parseDidUri } from '../../utilities/did/did';

import { Identity } from '../../utilities/identities/types';
import { usePopupData } from '../../utilities/popups/usePopupData';
import { IdentitiesCarousel } from '../../components/IdentitiesCarousel/IdentitiesCarousel';

import { backgroundCreateDidChannel } from '../../channels/CreateDidChannels/backgroundCreateDidChannel';
import { getIdentityCryptoFromSeed } from '../../utilities/identities/identities';

interface Props {
  identity: Identity;
}

export function CreateDidDApp({ identity }: Props): JSX.Element {
  const t = browser.i18n.getMessage;

  const { did } = identity;

  const { origin, submitter } = usePopupData<CreateDidOriginInput>();

  const passwordField = usePasswordField();

  const error = [
    isFullDid(did) && t('view_CreateDidDApp_error_full_did'),
    !did && t('view_CreateDidDApp_error_did_removed'),
  ].filter(Boolean)[0];

  const handleSubmit = useCallback(
    async (event: FormEvent) => {
      event.preventDefault();

      const { seed } = await passwordField.get(event);

      const { didDocument, sign } = await getIdentityCryptoFromSeed(seed);
      const { fullDid: did } = parseDidUri(didDocument.uri);

      const storeTx = await Did.getStoreTx(
        didDocument,
        submitter,
        async (input) => sign({ ...input, did }),
      );
      const signedExtrinsic = storeTx.method.toHex();

      await backgroundCreateDidChannel.return({ signedExtrinsic });
      window.close();
    },
    [passwordField, submitter],
  );

  const handleCancel = useCallback(async () => {
    await backgroundCreateDidChannel.throw('Rejected');
    window.close();
  }, []);

  return (
    <form className={styles.container} onSubmit={handleSubmit}>
      <h1 className={styles.heading}>{t('view_CreateDidDApp_heading')}</h1>

      <IdentitiesCarousel identity={identity} />

      <section className={styles.details}>
        <p className={styles.label}>{t('view_CreateDidDApp_origin')}</p>
        <p className={styles.origin}>{origin}</p>
      </section>

      <PasswordField identity={identity} autoFocus password={passwordField} />

      <p className={styles.buttonsLine}>
        <button onClick={handleCancel} type="button" className={styles.reject}>
          {t('common_action_cancel')}
        </button>
        <button
          type="submit"
          className={styles.submit}
          disabled={Boolean(error)}
        >
          {t('common_action_sign')}
        </button>
        <output className={styles.errorTooltip} hidden={!error}>
          {error}
        </output>
      </p>
    </form>
  );
}
