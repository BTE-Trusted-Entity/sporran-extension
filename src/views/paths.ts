export { generatePath } from 'react-router-dom';

export const paths = {
  home: '/',
  settings: '/settings',
  access: '/access',
  identity: {
    base: '/identity/',
    add: '/identity/add',
    create: {
      start: '/identity/create',
      backup: '/identity/create/backup',
      verify: '/identity/create/verify',
      password: '/identity/create/password',
    },
    import: {
      start: '/identity/import',
      password: '/identity/import/password',
    },

    overview: '/identity/:address/:type(created|imported|pwreset)?',
    reset: {
      start: '/identity/:address/reset',
      password: '/identity/:address/reset/password',
    },
    send: {
      start: '/identity/:address/send',
      review: '/identity/:address/send/review',
      warning: '/identity/:address/send/warning',
    },
    receive: '/identity/:address/receive',
    credentials: {
      base: '/identity/:address/credentials',
      presentation: '/identity/:address/credentials/:hash/presentation',
    },
    remove: '/identity/:address/remove',
    did: {
      upgrade: {
        start: '/identity/:address/did/upgrade',
        kilt: '/identity/:address/did/upgrade/kilt',
        euro: '/identity/:address/did/upgrade/euro',
      },
      manage: {
        start: '/identity/:address/did',
        downgrade: {
          base: '/identity/:address/did/downgrade',
          warning: {
            web3name: '/identity/:address/did/downgrade/warning/web3name',
            credentials: '/identity/:address/did/downgrade/warning/credentials',
          },
          sign: '/identity/:address/did/downgrade/sign',
        },
        endpoints: {
          start: '/identity/:address/did/endpoints',
          edit: '/identity/:address/did/endpoints/:id?',
          sign: '/identity/:address/did/endpoints/sign',
        },
      },
      repair: '/identity/:address/did/repair',
    },
    web3name: {
      create: {
        base: '/identity/:address/web3name/create',
        info: '/identity/:address/web3name/create/info',
        form: '/identity/:address/web3name/create/form',
        choose: '/identity/:address/web3name/create/choose/:web3name',
        kilt: '/identity/:address/web3name/create/kilt/:web3name',
        euro: '/identity/:address/web3name/create/euro/:web3name',
      },
      manage: {
        start: '/identity/:address/web3name',
        remove: '/identity/:address/web3name/remove',
      },
    },
  },
  popup: {
    base: '/popup',
    import: '/popup/import',
    access: '/popup/access',
    claim: '/identity/:address/claim',
    save: '/popup/save',
    share: {
      start: '/popup/share',
      sign: '/popup/share/sign',
    },
    sign: '/popup/sign',
    signRaw: '/popup/signRaw',
    signDid: {
      start: '/identity/:address/did/sign',
      sign: '/identity/:address/did/sign/password',
      credentials: '/identity/:address/did/sign/credentials',
    },
    signDidExtrinsic: '/identity/:address/did/signExtrinsic',
    createDid: '/identity/:address/did/create',
  },
};
