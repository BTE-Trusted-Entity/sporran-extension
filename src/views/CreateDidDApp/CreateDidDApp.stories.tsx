import { Meta } from '@storybook/react';

import { PopupTestProvider } from '../../utilities/popups/PopupTestProvider';
import { CreateDidOriginInput } from '../../channels/CreateDidChannels/types';

import { moreIdentitiesMock as identities } from '../../utilities/identities/IdentitiesProvider.mock';
import { paths } from '../paths';

import { CreateDidDApp } from './CreateDidDApp';

export default {
  title: 'Views/CreateDidDApp',
  component: CreateDidDApp,
} as Meta;

const input: CreateDidOriginInput = {
  dAppName: 'dApp',
  origin: 'https://example.org/foo',
  submitter: '4pUVoTJ69JMuapNducHJPU68nGkQXB7R9xAWY9dmvUh42653',
};

export function LightDid(): JSX.Element {
  return (
    <PopupTestProvider path={paths.popup.createDid} data={input}>
      <CreateDidDApp
        identity={
          identities['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1']
        }
      />
    </PopupTestProvider>
  );
}

export function FullDid(): JSX.Element {
  return (
    <PopupTestProvider path={paths.popup.createDid} data={input}>
      <CreateDidDApp
        identity={
          identities['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo']
        }
      />
    </PopupTestProvider>
  );
}

export function NoDid(): JSX.Element {
  return (
    <PopupTestProvider path={paths.popup.createDid} data={input}>
      <CreateDidDApp
        identity={
          identities['4rZ7pGtvmLhAYesf7DAzLQixdTEwWPN3emKb44bKVXqSoTZB']
        }
      />
    </PopupTestProvider>
  );
}
