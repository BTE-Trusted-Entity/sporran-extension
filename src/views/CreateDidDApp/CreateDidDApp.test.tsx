import { render } from '../../testing/testing';

import { PopupTestProvider } from '../../utilities/popups/PopupTestProvider';
import { paths } from '../paths';

import { waitForGetPassword } from '../../channels/SavedPasswordsChannels/SavedPasswordsChannels.mock';

import { moreIdentitiesMock as identities } from '../../utilities/identities/IdentitiesProvider.mock';

import { CreateDidOriginInput } from '../../channels/CreateDidChannels/types';

import { mockIsFullDid } from '../../utilities/did/did.mock';

import { CreateDidDApp } from './CreateDidDApp';

const mockInput: CreateDidOriginInput = {
  dAppName: 'dApp',
  origin: 'https://example.org/foo',
  submitter: '4pUVoTJ69JMuapNducHJPU68nGkQXB7R9xAWY9dmvUh42653',
};

describe('CreateDidDApp', () => {
  it('should render with light DID', async () => {
    const { container } = render(
      <PopupTestProvider path={paths.popup.createDid} data={mockInput}>
        <CreateDidDApp
          identity={
            identities['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1']
          }
        />
      </PopupTestProvider>,
    );
    await waitForGetPassword();

    expect(container).toMatchSnapshot();
  });

  it('should render with full DID', async () => {
    mockIsFullDid(true);

    const { container } = render(
      <PopupTestProvider path={paths.popup.createDid} data={mockInput}>
        <CreateDidDApp
          identity={
            identities['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo']
          }
        />
      </PopupTestProvider>,
    );
    await waitForGetPassword();

    expect(container).toMatchSnapshot();
  });

  it('should render with removed DID', async () => {
    mockIsFullDid(false);

    const { container } = render(
      <PopupTestProvider path={paths.popup.createDid} data={mockInput}>
        <CreateDidDApp
          identity={
            identities['4rZ7pGtvmLhAYesf7DAzLQixdTEwWPN3emKb44bKVXqSoTZB']
          }
        />
      </PopupTestProvider>,
    );
    await waitForGetPassword();

    expect(container).toMatchSnapshot();
  });
});
