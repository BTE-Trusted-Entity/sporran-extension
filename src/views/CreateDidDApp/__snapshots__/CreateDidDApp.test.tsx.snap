// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateDidDApp should render with full DID 1`] = `
<div>
  <form
    class="container"
  >
    <h1
      class="heading"
    >
      Create DID
    </h1>
    <div
      class="container"
    >
      <section
        class="container"
      >
        <div
          class="avatar fullDid"
        >
          Identicon
        </div>
        <div
          class="centeredNameLine"
        >
          <span
            class="name"
          >
            Identity with on-chain DID and lots of coins
          </span>
        </div>
      </section>
      <a
        aria-label="Light DID Identity"
        class="left"
        href="/?data=eyJkQXBwTmFtZSI6ImRBcHAiLCJvcmlnaW4iOiJodHRwczovL2V4YW1wbGUub3JnL2ZvbyIsInN1Ym1pdHRlciI6IjRwVVZvVEo2OUpNdWFwTmR1Y0hKUFU2OG5Ha1FYQjdSOXhBV1k5ZG12VWg0MjY1MyJ9"
        title="Light DID Identity"
      />
      <a
        aria-label="Identity which has a very very very very long name and might not even fit on the screen"
        class="right"
        href="/?data=eyJkQXBwTmFtZSI6ImRBcHAiLCJvcmlnaW4iOiJodHRwczovL2V4YW1wbGUub3JnL2ZvbyIsInN1Ym1pdHRlciI6IjRwVVZvVEo2OUpNdWFwTmR1Y0hKUFU2OG5Ha1FYQjdSOXhBV1k5ZG12VWg0MjY1MyJ9"
        title="Identity which has a very very very very long name and might not even fit on the screen"
      />
      <ul
        class="bubbles"
      >
        <li
          class="item"
        >
          <a
            aria-current="page"
            aria-label="Light DID Identity"
            class="bubble bubbleActive"
            href="/?data=eyJkQXBwTmFtZSI6ImRBcHAiLCJvcmlnaW4iOiJodHRwczovL2V4YW1wbGUub3JnL2ZvbyIsInN1Ym1pdHRlciI6IjRwVVZvVEo2OUpNdWFwTmR1Y0hKUFU2OG5Ha1FYQjdSOXhBV1k5ZG12VWg0MjY1MyJ9"
            title="Light DID Identity"
          />
        </li>
        <li
          class="item"
        >
          <a
            aria-current="page"
            aria-label="Identity with on-chain DID and lots of coins"
            class="bubble bubbleActive"
            href="/?data=eyJkQXBwTmFtZSI6ImRBcHAiLCJvcmlnaW4iOiJodHRwczovL2V4YW1wbGUub3JnL2ZvbyIsInN1Ym1pdHRlciI6IjRwVVZvVEo2OUpNdWFwTmR1Y0hKUFU2OG5Ha1FYQjdSOXhBV1k5ZG12VWg0MjY1MyJ9"
            title="Identity with on-chain DID and lots of coins"
          />
        </li>
        <li
          class="item"
        >
          <a
            aria-current="page"
            aria-label="Identity which has a very very very very long name and might not even fit on the screen"
            class="bubble bubbleActive"
            href="/?data=eyJkQXBwTmFtZSI6ImRBcHAiLCJvcmlnaW4iOiJodHRwczovL2V4YW1wbGUub3JnL2ZvbyIsInN1Ym1pdHRlciI6IjRwVVZvVEo2OUpNdWFwTmR1Y0hKUFU2OG5Ha1FYQjdSOXhBV1k5ZG12VWg0MjY1MyJ9"
            title="Identity which has a very very very very long name and might not even fit on the screen"
          />
        </li>
        <li
          class="item"
        >
          <a
            aria-current="page"
            aria-label="Add Identity"
            class="add addActive"
            href="/"
            title="Add Identity"
          />
        </li>
      </ul>
    </div>
    <section
      class="details"
    >
      <p
        class="label"
      >
        Request from:
      </p>
      <p
        class="origin"
      >
        https://example.org/foo
      </p>
    </section>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
    <p
      class="buttonsLine"
    >
      <button
        class="reject"
        type="button"
      >
        Cancel
      </button>
      <button
        class="submit"
        disabled=""
        type="submit"
      >
        Sign
      </button>
      <output
        class="errorTooltip"
      >
        DID already exists
      </output>
    </p>
  </form>
</div>
`;

exports[`CreateDidDApp should render with light DID 1`] = `
<div>
  <form
    class="container"
  >
    <h1
      class="heading"
    >
      Create DID
    </h1>
    <div
      class="container"
    >
      <section
        class="container"
      >
        <div
          class="avatar"
        >
          Identicon
        </div>
        <div
          class="centeredNameLine"
        >
          <span
            class="name"
          >
            Light DID Identity
          </span>
        </div>
      </section>
      <a
        aria-label="Add Identity"
        class="left"
        href="/?data=eyJkQXBwTmFtZSI6ImRBcHAiLCJvcmlnaW4iOiJodHRwczovL2V4YW1wbGUub3JnL2ZvbyIsInN1Ym1pdHRlciI6IjRwVVZvVEo2OUpNdWFwTmR1Y0hKUFU2OG5Ha1FYQjdSOXhBV1k5ZG12VWg0MjY1MyJ9"
        title="Add Identity"
      />
      <a
        aria-label="Identity with on-chain DID and lots of coins"
        class="right"
        href="/?data=eyJkQXBwTmFtZSI6ImRBcHAiLCJvcmlnaW4iOiJodHRwczovL2V4YW1wbGUub3JnL2ZvbyIsInN1Ym1pdHRlciI6IjRwVVZvVEo2OUpNdWFwTmR1Y0hKUFU2OG5Ha1FYQjdSOXhBV1k5ZG12VWg0MjY1MyJ9"
        title="Identity with on-chain DID and lots of coins"
      />
      <ul
        class="bubbles"
      >
        <li
          class="item"
        >
          <a
            aria-current="page"
            aria-label="Light DID Identity"
            class="bubble bubbleActive"
            href="/?data=eyJkQXBwTmFtZSI6ImRBcHAiLCJvcmlnaW4iOiJodHRwczovL2V4YW1wbGUub3JnL2ZvbyIsInN1Ym1pdHRlciI6IjRwVVZvVEo2OUpNdWFwTmR1Y0hKUFU2OG5Ha1FYQjdSOXhBV1k5ZG12VWg0MjY1MyJ9"
            title="Light DID Identity"
          />
        </li>
        <li
          class="item"
        >
          <a
            aria-current="page"
            aria-label="Identity with on-chain DID and lots of coins"
            class="bubble bubbleActive"
            href="/?data=eyJkQXBwTmFtZSI6ImRBcHAiLCJvcmlnaW4iOiJodHRwczovL2V4YW1wbGUub3JnL2ZvbyIsInN1Ym1pdHRlciI6IjRwVVZvVEo2OUpNdWFwTmR1Y0hKUFU2OG5Ha1FYQjdSOXhBV1k5ZG12VWg0MjY1MyJ9"
            title="Identity with on-chain DID and lots of coins"
          />
        </li>
        <li
          class="item"
        >
          <a
            aria-current="page"
            aria-label="Identity which has a very very very very long name and might not even fit on the screen"
            class="bubble bubbleActive"
            href="/?data=eyJkQXBwTmFtZSI6ImRBcHAiLCJvcmlnaW4iOiJodHRwczovL2V4YW1wbGUub3JnL2ZvbyIsInN1Ym1pdHRlciI6IjRwVVZvVEo2OUpNdWFwTmR1Y0hKUFU2OG5Ha1FYQjdSOXhBV1k5ZG12VWg0MjY1MyJ9"
            title="Identity which has a very very very very long name and might not even fit on the screen"
          />
        </li>
        <li
          class="item"
        >
          <a
            aria-current="page"
            aria-label="Add Identity"
            class="add addActive"
            href="/"
            title="Add Identity"
          />
        </li>
      </ul>
    </div>
    <section
      class="details"
    >
      <p
        class="label"
      >
        Request from:
      </p>
      <p
        class="origin"
      >
        https://example.org/foo
      </p>
    </section>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
    <p
      class="buttonsLine"
    >
      <button
        class="reject"
        type="button"
      >
        Cancel
      </button>
      <button
        class="submit"
        type="submit"
      >
        Sign
      </button>
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
  </form>
</div>
`;

exports[`CreateDidDApp should render with removed DID 1`] = `
<div>
  <form
    class="container"
  >
    <h1
      class="heading"
    >
      Create DID
    </h1>
    <div
      class="container"
    >
      <section
        class="container"
      >
        <div
          class="avatar"
        >
          Identicon
        </div>
        <div
          class="centeredNameLine"
        >
          <span
            class="name"
          >
            Identity with deleted on-chain DID
          </span>
        </div>
      </section>
      <a
        aria-label="Add Identity"
        class="left"
        href="/?data=eyJkQXBwTmFtZSI6ImRBcHAiLCJvcmlnaW4iOiJodHRwczovL2V4YW1wbGUub3JnL2ZvbyIsInN1Ym1pdHRlciI6IjRwVVZvVEo2OUpNdWFwTmR1Y0hKUFU2OG5Ha1FYQjdSOXhBV1k5ZG12VWg0MjY1MyJ9"
        title="Add Identity"
      />
      <a
        aria-label="Light DID Identity"
        class="right"
        href="/?data=eyJkQXBwTmFtZSI6ImRBcHAiLCJvcmlnaW4iOiJodHRwczovL2V4YW1wbGUub3JnL2ZvbyIsInN1Ym1pdHRlciI6IjRwVVZvVEo2OUpNdWFwTmR1Y0hKUFU2OG5Ha1FYQjdSOXhBV1k5ZG12VWg0MjY1MyJ9"
        title="Light DID Identity"
      />
      <ul
        class="bubbles"
      >
        <li
          class="item"
        >
          <a
            aria-current="page"
            aria-label="Light DID Identity"
            class="bubble bubbleActive"
            href="/?data=eyJkQXBwTmFtZSI6ImRBcHAiLCJvcmlnaW4iOiJodHRwczovL2V4YW1wbGUub3JnL2ZvbyIsInN1Ym1pdHRlciI6IjRwVVZvVEo2OUpNdWFwTmR1Y0hKUFU2OG5Ha1FYQjdSOXhBV1k5ZG12VWg0MjY1MyJ9"
            title="Light DID Identity"
          />
        </li>
        <li
          class="item"
        >
          <a
            aria-current="page"
            aria-label="Identity with on-chain DID and lots of coins"
            class="bubble bubbleActive"
            href="/?data=eyJkQXBwTmFtZSI6ImRBcHAiLCJvcmlnaW4iOiJodHRwczovL2V4YW1wbGUub3JnL2ZvbyIsInN1Ym1pdHRlciI6IjRwVVZvVEo2OUpNdWFwTmR1Y0hKUFU2OG5Ha1FYQjdSOXhBV1k5ZG12VWg0MjY1MyJ9"
            title="Identity with on-chain DID and lots of coins"
          />
        </li>
        <li
          class="item"
        >
          <a
            aria-current="page"
            aria-label="Identity which has a very very very very long name and might not even fit on the screen"
            class="bubble bubbleActive"
            href="/?data=eyJkQXBwTmFtZSI6ImRBcHAiLCJvcmlnaW4iOiJodHRwczovL2V4YW1wbGUub3JnL2ZvbyIsInN1Ym1pdHRlciI6IjRwVVZvVEo2OUpNdWFwTmR1Y0hKUFU2OG5Ha1FYQjdSOXhBV1k5ZG12VWg0MjY1MyJ9"
            title="Identity which has a very very very very long name and might not even fit on the screen"
          />
        </li>
        <li
          class="item"
        >
          <a
            aria-current="page"
            aria-label="Add Identity"
            class="add addActive"
            href="/"
            title="Add Identity"
          />
        </li>
      </ul>
    </div>
    <section
      class="details"
    >
      <p
        class="label"
      >
        Request from:
      </p>
      <p
        class="origin"
      >
        https://example.org/foo
      </p>
    </section>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4rZ7pGtvmLhAYesf7DAzLQixdTEwWPN3emKb44bKVXqSoTZB/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
    <p
      class="buttonsLine"
    >
      <button
        class="reject"
        type="button"
      >
        Cancel
      </button>
      <button
        class="submit"
        disabled=""
        type="submit"
      >
        Sign
      </button>
      <output
        class="errorTooltip"
      >
        DID already removed
      </output>
    </p>
  </form>
</div>
`;
